crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -172,6 +181,91 @@ impl World {
         }
         Ok(())
     }
+
+    /// Run a system one-shot by itself.
+    /// It calls like [`RunSystemOnce::run_system_once`](crate::system::RunSystemOnce::run_system_once),
+    /// but behaves like [`World::run_system`].
+    /// System will be registered on first run, then its state is kept for after calls.
+    ///
+    /// # Limitations
+    ///
+    /// See [`World::run_system`].
+    ///
+    /// # Attention
+    ///
+    /// Please note that closure is identified by its **occurrence** no matter what it captures,
+    /// thus this code will call the first cached one rather than actual passed one.
+    ///
+    /// That means, codes like below will not work as expected.
+    ///
+    /// ```rust
+    /// use bevy_ecs::prelude::World;
+    /// let mut world = World::new();
+    /// // This will creates closures regarded as one same closure
+    /// let make_system = |n| move || println!("{n}");
+    /// let _ = world.run_system_singleton(make_system(0)); // prints 0
+    /// let _ = world.run_system_singleton(make_system(1)); // prints 0 same
+    /// ```
+    pub fn run_system_singleton<M, T: IntoSystem<(), (), M>>(
+        &mut self,
+        system: T,
+    ) -> Result<(), UnregisteredSystemError> {
+        // create the hashmap if it doesn't exist
+        let mut system_map = self.get_resource_or_insert_with(|| SystemMap::default());
+
+        let system = IntoSystem::<(), (), M>::into_system(system);
+
+        // check captures of closure
+        if std::mem::size_of::<T>() != 0 {
+            info!(target: "run_system_singleton", "Closure with capture(s) runs, be sure that you're not relying on differently captured versions of one closure.")
+        }
+        // use the system type id as the key
+        let system_type_id = system.type_id();
+        // take system out
+        let UnregisteredSystem {
+            mut initialized,
+            system,
+        } = system_map
+            .map
+            .remove(&system_type_id)
+            .unwrap_or_else(|| UnregisteredSystem {
+                initialized: false,
+                system: Some(Box::new(system)),
+            });
+        // insert the marker
+        system_map.map.insert(
+            system_type_id,
+            UnregisteredSystem {
+                initialized: false,
+                system: None,
+            },
+        );
+        // check if runs recursively
+        let Some(mut system) = system else {
+            return Err(UnregisteredSystemError::Recursive);
+        };
+
+        // run the system
+        if !initialized {
+            system.initialize(self);
+            initialized = true;
+        }
```

Comment:
This is currently `removing` and re-`insert`ing into the systems map, which feels pretty weird IMO since the intended meaning is more like to remove only the actual system inside the `system` `Option` field. This also requires two lookups which are not ideal. You could consider using `entry` to get/initialize the entry in the map and then `take` on the `system` field to replace it with `None`.

Something like this (not tested, probably needs changes to the final `insert` as well):

```suggestion
        let UnregisteredSystem {
            initialized,
            system,
        } = system_map
            .map
            .entry(system_type_id)
            .or_insert_with(|| UnregisteredSystem {
                initialized: false,
                system: Some(Box::new(system)),
            });
        // check if runs recursively
        let Some(mut system) = system.take() else {
            return Err(UnregisteredSystemError::Recursive);
        };

        // run the system
        if !*initialized {
            system.initialize(self);
            *initialized = true;
        }
```